### ISSUE TITLE
[python] Schema evolution for `obs` and `var`

### ISSUE THREAD

**Issue by royassis** [single-cell-data/TileDB-SOMA#499]
Hey 

I'm trying to update the obs of a soma dataset and i'm getting an error:

```python
Traceback (most recent call last):
  File "/homefolder/roya/anaconda3/envs/soma/lib/python3.9/site-packages/IPython/core/interactiveshell.py", line 3398, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-50-80115732a5ae>", line 1, in <cell line: 1>
    from_anndata_update_obs_and_var(soma, adata_plus, _unused="")
  File "/homefolder/roya/anaconda3/envs/soma/lib/python3.9/site-packages/tiledbsoma/io.py", line 286, in from_anndata_update_obs_and_var
    soma.obs.from_dataframe(dataframe=anndata.obs, extent=256)
  File "/homefolder/roya/anaconda3/envs/soma/lib/python3.9/site-packages/tiledbsoma/annotation_dataframe.py", line 411, in from_dataframe
    tiledb.from_pandas(
  File "/homefolder/roya/anaconda3/envs/soma/lib/python3.9/site-packages/tiledb/dataframe_.py", line 450, in from_pandas
    _from_pandas(uri, dataframe, tiledb_args)
  File "/homefolder/roya/anaconda3/envs/soma/lib/python3.9/site-packages/tiledb/dataframe_.py", line 522, in _from_pandas
    _write_array(
  File "/homefolder/roya/anaconda3/envs/soma/lib/python3.9/site-packages/tiledb/dataframe_.py", line 606, in _write_array
    libtiledb._setitem_impl_sparse(A, tuple(coords), write_dict, nullmaps)
  File "tiledb/libtiledb.pyx", line 5017, in tiledb.libtiledb._setitem_impl_sparse
tiledb.cc.TileDBError: Sparse write input data count does not match number of attributes
```

code to replicate:
```
import anndata
import pandas as pd
import numpy as np
import tiledbsoma
from tiledbsoma.io import from_anndata_update_obs_and_var, from_anndata

X = np.arange(9).reshape(3, 3)
obs = pd.DataFrame(index=["a", "b", "c"])
var = pd.DataFrame(index=["a", "b", "c"])

adata = anndata.AnnData(X=X, obs=obs, var=var)

new_adata = adata.copy()
new_obs = obs
new_obs["a"] = 1
new_adata.obs = new_obs

soma = tiledbsoma.SOMA("/tmp/test")

from_anndata(soma, adata)

from_anndata_update_obs_and_var(soma, new_adata, _unused="")
```

**Comment by johnkerl** [comment]
@royassis thanks for the report!

The issue happens when there is a schema change:

```
OLD OBS
Empty DataFrame
Columns: []
Index: [a, b, c]

NEW OBS
   a
a  1
b  1
c  1
```

There are a few levels of things we can do here:

* Fail the second write with a more intuitive error message
* Remove and replace the `obs`
* Leverage TileDB's "schema evolution" feature

I'll do a bit more research and get back to you.

Thanks again for the report! :)

**Comment by johnkerl** [comment]
@royassis this is implemented in TileDB-SOMA 1.5.0 -- e.g. `tiledbsoma.io.update_obs` and `tiledbsoma.io.update_var`.

That upcoming release is not yet tagged, so I'll keep this issue open until then -- as well, I'll link to a notebook example.

**Comment by johnkerl** [comment]
@royassis this is available here
https://github.com/single-cell-data/TileDB-SOMA/releases/tag/1.5.0rc1
-- the upcoming 1.5.0 won't affect this so this is ready for eval  now.

If you don't mind I'll go ahead and close this -- and if you have unanswered concerns or I missed something, feel free to re-open?
